### ISSUE TITLE
Select a specific branch in webhook configuration on push

### ISSUE THREAD

**Issue by Womps** [gogits/gogs#4105]
- Gogs version (or commit ref): 0.9.135.0208
- Git version: 1.7.1
- Operating system: Linux CentOs 6.8
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gogs.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
Hello,

We are working on a platform which include jenkins and gogs. 
We installed the plugin for gogs in jenkins. And it works very well between gogs and jenkins. 
We configured WebHooks in gogs, to trigger a build in jenkins, when a new commit has been pushed to the repository. 

For a project, we have 2 branches : 
   - 1 master branch which is our "release" branch
   - 1 develop branch, which is used to work on next releases.

For each branch exists a job build in jenkins. So, we have a build for develop branch, and a build for the master branch.

Now, we have a problem. When we push on "develop" branch on git, the jenkins master job is also build... The thing is we don't want to build the jenkins master job every time a developer push to the repository... We want that when we push on the develop branch, only the develop job is triggered in jenkins. 

Actually, we developed a git hook script to trigger the job corresponding to the branch that has been pushed, but what about developping a new feature in gogs that allows repositories administrators to select the branch(es) on which the push request has been executed, when we configure the webhooks ?

**Comment by Unknwon** [comment]
Hi, I think Gogs sends out which branch code was pushed on, does Jenkins ignore that?

**Comment by Womps** [comment]
I think jenkins don't ignore anything, it's just gogs which trigger our 2 webhooks. It works well, but take a look :
On our project repository, we configured 2 webhooks. 1 For master job and 1 for develop job : 
![webhooks_master-develop](https://cloud.githubusercontent.com/assets/12201026/22791237/fab399c0-eee8-11e6-84d7-ca3044929e34.png)

This is how the develop webhook is configured. The master webhook is configured with exactly the same parameters, except the URI with MASTER, instead of DEVELOP
![develop_configure_webhook](https://cloud.githubusercontent.com/assets/12201026/22791317/421b96a0-eee9-11e6-99d0-d79cd014a246.png)

I think i was not clear in my first post : When a developer push on develop branch in git, gogs trigger the 2 webhooks, so the 2 job are triggered in jenkins.

**Comment by Unknwon** [comment]
I see!

**Comment by Liangdi** [comment]
确实，指定分支的事件推送，这个可以满足更灵活的部署配置，我们也有这样的需求

**Comment by Unknwon** [comment]
How to handle cases when a webhook event is not related to a branch? Say issues?

**Comment by clement-garrigou** [comment]
> Actually, we developed a git hook script to trigger the job corresponding to the branch that has been pushed
@Womps Can you explain how this script works ? Or share this script ? Thanks !

**Comment by Womps** [comment]
> How to handle cases when a webhook event is not related to a branch? Say issues?

@Unknwon Maybe we can have a list with repo's branches, and select the branch that we need. And if the webhook doesn't need to be related to a branch, maybe we can have an option "Disable", or "Global" ?

@clement-garrigou The script is in post-receive scripts. We you use these scripts, you got some parameters, so we use regex in parameters to match the name of the branch that have been pushed. And from the name of the branch, we trigger a job in jenkins, using curl command, to POST a request to different URI of jenkins. We used this documentation to use curl command : https://wiki.jenkins-ci.org/display/JENKINS/Remote+access+API

**Comment by Unknwon** [comment]
My thought is when you choose "only push event" then you're able to specify a branch. How's that sound? Otherwise branch options just isn't appropriate.

**Comment by Womps** [comment]
Currently, when you check the option "Only push event", there is no choice, if a push is done on develop branch, then the webhook is triggered.

What we ask is when you check the option "Only push event", then ask to the user to specify on which branch the push must be done, to trigger the webhook. So you're able to specify a branch that exist on the gogs repo. And it can be a branch, OR maybe we can have an option selected by default, so that it can be a "Global" webhook, which can be common to all branches on the repo.

**Comment by Unknwon** [comment]
What I am talking is "specify a branch" is only making sense when user choose "only push event". This is a proposal of implementation, I never said it's already implemented.

**Comment by Womps** [comment]
Yes yes this is what we thought, specify a branch when you choose only push event

**Comment by Unknwon** [comment]
Great, sounds good.

**Comment by xxlaila** [comment]
A github library inside a number of projects, jenkins for each project configuration of a job, gogs plug-in configuration, submitted a project code to trigger Ajob, gogs hook will only trigger A job will not trigger other job The

**Comment by BoZhaoUTSC** [comment]
This feature will be incredibly useful.

**Comment by ecs-hk** [comment]
Agreed, this is useful for Gogs' Webhook push event + Jenkins' "Trigger Builds Remotely" option.

[Jenkins wiki - Allowing Developers to do Forced Builds with Security On](https://wiki.jenkins.io/display/JENKINS/Quick+and+Simple+Security)

When I push to develop I'd like a Jenkins build job to run. When I push to master I'd like a separate Jenkins build job to run. The URI (with auth token) Jenkins uses doesn't appear to have a lot of intelligence behind it as of this writing. If it receives a specific HTTP request (unique to a specific build job) it will run the build job.

**Comment by forpix** [comment]
I have a  job in Jenkins is configured to trigger through web-hook. In GitHub, I have a repository with two branches 1. Master 2.Working, now I have committed/created a new file  on Working branch and Job is triggered, which exactly I want to have.😊

Now the committed file in Working branch should have to be merge to Master so PR should be needed, which leads to trigger the job again.
(Because of PR from Working to Master again new build is triggered which I don't want..😐)
any specific branch web-hook available in GitHub

**Comment by Manu-H** [comment]
Anything new here? Would be happy with this feature 👍

**Comment by jazz-mobility** [comment]
Any timelines when can we have this? ETA would be great here.
Keep up the good work. :)

**Comment by shaunwarman** [comment]
A common practice is to use the webhook for when releases are published to the repository. So if you use a `develop` / `release` like branching strategy, cut a release with a version and that release will trigger a webhook.

**Comment by forpix** [comment]
Dear Team,

Is there a way to create a Pull Request from command line terminal (git)

**Comment by zkdm** [comment]
Is this feature still on the roadmap?

**Comment by AcerFeng** [comment]
两年了，这个功能能不能实现呢？或者在 post 信息中给个当前触发的分支信息也行，我们可以自己去判断

**Comment by sunilkarthy89** [comment]
You can use refs:
In Github hooks:
* Payload URL: https://jenkins.example.com/generic-webhook-trigger/invoke?token=<Any_token_value>
* content type should be : application/json

In jenkins:
* Enable generic-webhook-trigger and add post content parameters - variable "ref" and expression "$.ref".
* Token value from the <Any_token_value> in payload url
* Go to "Optional filter"
* Expression should be "^(refs/heads/release)"
* For multiple branches use "^(refs/heads/develop|refs/heads/release|refs/heads/master)"
* For feature branches use "^(refs/heads/feature/.+)$"
* You can use mix of all as well
* Text should be $ref
* enable poll scm as well

This should trigger the jenkins only for the particular branch you are using the filter expression.

**Comment by SpaceNinjaApe** [comment]
In addition to @sunilkarthy89 solution you could make your Job a parameterized job and use a parameter-name as value for the generic-webhook token-value.

After that, you could add conditional handling (via [plugin](https://wiki.jenkins.io/display/JENKINS/Conditional+BuildStep+Plugin) or you can handle the variable in shell `if [ "{$param} = "foo" ]; then`.
This way you can do different stuff for different branches in one job. Keeps your Jenkins joblist clean.

**Comment by geekfred** [comment]
> What I am talking is "specify a branch" is only making sense when user choose "only push event". This is a proposal of implementation, I never said it's already implemented.

I know this is an old thread, but I wanted to let you know what I did since I was having the same problems regarding Webhooks for specific branches. It turns out that AWS CodeCommit and Azure Devops both let you set up Webhooks per branches, so what I did was to hook up one of this services as extra remotes and it works great, I keep using my GitHub repo as my main repo but if I work on the develop branch it gets pushed to every remote but it is Azure Devops the one that fires up Webhooks depending on which branch it gets the code pushed to. Works like a charm.

**Comment by oiricaud** [comment]
any ideas how you can do this with tekton?

**Comment by metka495** [comment]
> 确实，指定分支的事件推送，这个可以满足更灵活的部署配置，我们也有这样的需求

我现在也卡在这边了 ，每次 dev分支推送  master分支就会莫名其妙的构建 好烦阿，新版本的gogs不知道有没有更新

**Comment by chetansheth** [comment]
I am not finding option to choose branch after selecting "Just the push event." Is there any idea or other options where I can specify the Branch name ?

**Comment by yvolodym** [comment]
You can enter branch as an option e.g
http://jenkins:8080/gogs-webhook/?job=msnetwork/master&branch=master

Am Mo., 20. Dez. 2021 um 15:29 Uhr schrieb chetansheth <
***@***.***>:
